integrations/parquet/src/async_reader.rs

```diff
@@ -0,0 +1,343 @@
+
+/// AsyncReader implements AsyncFileReader trait by using opendal.
+///
+/// ```no_run
+/// use std::sync::Arc;
+/// use arrow::array::{ArrayRef, Int64Array, RecordBatch};
+///
+/// use futures::StreamExt;
+/// use opendal::{services::S3Config, Operator};
+/// use parquet::arrow::{AsyncArrowWriter, ParquetRecordBatchStreamBuilder};
+/// use parquet_opendal::{AsyncReader, AsyncWriter};
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let mut cfg = S3Config::default();
+///     cfg.access_key_id = Some("my_access_key".to_string());
+///     cfg.secret_access_key = Some("my_secret_key".to_string());
+///     cfg.endpoint = Some("my_endpoint".to_string());
+///     cfg.region = Some("my_region".to_string());
+///     cfg.bucket = "my_bucket".to_string();
+///
+///     // Create a new operator
+///     let operator = Operator::from_config(cfg).unwrap().finish();
+///     let path = "/path/to/file.parquet";
+///
+///     // Create an async writer
+///     let writer = AsyncWriter::new(
+///         operator
+///             .writer_with(path)
+///             .chunk(32 * 1024 * 1024)
+///             .concurrent(8)
+///             .await
+///             .unwrap(),
+///     );
+///
+///     let col = Arc::new(Int64Array::from_iter_values([1, 2, 3])) as ArrayRef;
+///     let to_write = RecordBatch::try_from_iter([("col", col)]).unwrap();
+///     let mut writer = AsyncArrowWriter::try_new(writer, to_write.schema(), None).unwrap();
+///     writer.write(&to_write).await.unwrap();
+///     writer.close().await.unwrap();
+///
+///     // gap: Allow the underlying reader to merge small IOs
+///     // when the gap between multiple IO ranges is less than the threshold.
+///     let reader = operator
+///         .reader_with(path)
+///         .gap(512 * 1024)
+///         .chunk(16 * 1024 * 1024)
+///         .concurrent(16)
+///         .await
+///         .unwrap();
+///     let content_len = operator.stat(path).await.unwrap().content_length();
+///     let reader = AsyncReader::new(reader, content_len).with_footer_size(512 * 1024);
+///     let mut stream = ParquetRecordBatchStreamBuilder::new(reader)
+///         .await
+///         .unwrap()
+///         .build()
+///         .unwrap();
+///     let read = stream.next().await.unwrap().unwrap();
+///     assert_eq!(to_write, read);
+/// }
+/// ```
+pub struct AsyncReader {
+    inner: Reader,
+    content_len: u64,
+    // The prefetch size for fetching file footer.
+    footer_size: usize,
+}
+
+fn set_footer_size(footer_size: usize, content_size: u64) -> usize {
+    let footer_size = max(footer_size, FOOTER_SIZE);
+    min(footer_size as u64, content_size) as usize
+}
+
+impl AsyncReader {
+    /// Create a [`AsyncReader`] by given [`Reader`].
+    pub fn new(reader: Reader, content_len: u64) -> Self {
+        Self {
+            inner: reader,
+            content_len,
+            footer_size: set_footer_size(PREFETCH_FOOTER_SIZE, content_len),
```

Comment:
The current `footer_size` behavior is not what I expected. It seems to confuse the exact `footer_size` with `prefetch_footer_size`. The implementation aligns with the latter, but the naming suggests the former.

I suggest to use `prefetch_footer_size` if that's what we want. And an exact `footer_size` maybe useful too, I'm not sure.

Reply:
Renamed it to `prefetch_footer_size`